lib/cloud-operation/communication/communication-channel-base.ts

```diff
@@ -0,0 +1,120 @@
+import { EventEmitter } from "events";
+import { CloudOperationMessageTypes, CloudCommunicationChannelExitCodes, CloudCommunicationEvents } from "../../constants";
+
+export abstract class CommunicationChannelBase extends EventEmitter implements ICloudCommunicationChannel {
+	private static readonly PING_INTERVAL: number = 10000;
+	private static readonly MISSING_PING_REPLIES_COUNT: number = 6;
+
+	protected isConnected: boolean;
+	private pingInterval: NodeJS.Timer;
+	private echoSequenceNumber: number = 0;
+	private echoReplySequenceNumber: number = 0;
+
+	private handshakeCompleteResolve: (value?: PromiseLike<void>) => void;
+
+	constructor(protected cloudOperationId: string,
+		protected data: ICloudCommunicationChannelData<any>,
+		protected $logger: ILogger) {
+		super();
+	}
+
+	public async connect(): Promise<void> {
+		if (!this.isConnected) {
+			await this.connectCore();
+			this.isConnected = true;
+			await this.handshake();
+			this.startPing();
+		}
+	}
+
+	public async sendMessage<T>(message: ICloudOperationMessage<T>): Promise<string> {
+		if (this.isConnected) {
+			const result = await this.sendMessageCore(message);
+
+			return result;
+		}
+
+		throw new Error("Communication channel not connected");
+	}
+
+	public async close(code: number, reason?: string): Promise<void> {
+		if (!this.isConnected) {
+			return;
+		}
+
+		try {
+			if (this.pingInterval) {
+				clearInterval(this.pingInterval);
+				this.pingInterval = null;
+			}
+
+			const stopMsg: ICloudOperationMessage<ICloudOperationStop> = {
+				type: CloudOperationMessageTypes.CLOUD_OPERATION_STOP,
+				cloudOperationId: this.cloudOperationId,
+				body: {
+					code,
+					reason
+				}
+			};
+			await this.sendMessage<ICloudOperationStop>(stopMsg);
+			await this.closeCore(code, reason);
+		} catch (err) {
+			this.$logger.trace(err);
```

Comment:
Why do we need the try .. catch here?

Reply:
Because we may be in a situation where we can't send stop message and we don't want to stop any other cleanup tasks.